doc/write-a-wasm-extension-with-cpp.md

```diff
@@ -0,0 +1,178 @@
+
+Apply the following [EnvoyFilter](../example/config/storage-cluster.yaml) to register the blob service cluster for Extension downloading. Here we use Google Cloud Storage as an example:
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: EnvoyFilter
+metadata:
+  name: example-cluster
+spec:
+  configPatches:
+  - applyTo: CLUSTER
+    match:
+      context: SIDECAR_INBOUND
+    patch:
+      operation: ADD
+      value: # cluster specification
+        name: google-storage
+        dns_lookup_family: V4_ONLY
+        type: STRICT_DNS
+        connect_timeout: 5s
+        transport_socket:
+          name: envoy.transport_sockets.tls
+        load_assignment:
+          cluster_name: storage.googleapis.com
+          endpoints:
+          - lb_endpoints:
+            - endpoint:
+                address:
+                  socket_address:
+                    address: storage.googleapis.com
+                    port_value: 443
+```
+
+Then apply the following [EnvoyFilter](../example/config/example-filter.yaml) to inject the extension into inbound sidecar listeners:
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: EnvoyFilter
+metadata:
+  name: example-filter
+spec:
+  configPatches:
+  - applyTo: HTTP_FILTER
+    match:
+      context: SIDECAR_INBOUND
+      listener:
+        filterChain:
+          filter:
+            name: envoy.http_connection_manager
+            subFilter:
+              name: envoy.router
+    patch:
+      operation: INSERT_BEFORE
+      value:
+        name: example-filter
+        typed_config:
+          '@type': type.googleapis.com/udpa.type.v1.TypedStruct
+          type_url: type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm
+          value:
+            config:
+              vm_config:
+                code:
+                  remote:
+                    http_uri:
+                      uri: https://storage.googleapis.com/istio-ecosystem/wasm-extensions/example.wasm
+                      cluster: google-storage
+                      timeout: 10s
+                    retry_policy:
+                      num_retries: 5
+                      retry_back_off:
+                        base_interval: 1s
+                        max_interval: 3s
+                    sha256: a96aacb4fa8fd81bb7a7112b1d46432ec1463b8142d320bc269533832aa3d9a7
+                runtime: envoy.wasm.runtime.v8
+```
+
+Now curl the `httpbin` service, and `x-custom-foo` should be added into the response!
+```
+$ curl -I 34.123.215.139
+HTTP/1.1 200 OK
+server: istio-envoy
+...
+x-wasm-custom: foo
+```
+
+## Step 7: Maintain your extension along with Istio releases
+
+After writing, testing, and deploying the extension successfully, it is time to make sure that your extension could evolve along with Istio. Currently it is recommended to build and deploy your extension following Istio releases. Specifically:
```

Comment:
Not for this PR but for later, I am thinking should this be a part of this repo? We maintain versions of this, similar to Istio versions, import istio e2e test framework too and users write istio integration test using their wasm filter and envoy patches. That way if a breaking change happens like api change in envoy filter or telemetry api, they would know about it immediately?

Reply:
Yeah what you described is the plan. Integration test is already done for basic_auth filter and should be required for any future extension. I am about to cut 1.8 release branch in this repo, and modules will be released and versioned as Istio.